ddl/index_cop.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/timeutil"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// copReadBatchFactor is the factor of batch size of coprocessor read.
+// It multiplies the tidb_ddl_reorg_batch_size to avoid sending too many cop requests for the same handle range.
+const copReadBatchFactor = 10
+
+func (w *addIndexWorker) fetchRowColValsFromCop(txn kv.Transaction, handleRange reorgBackfillTask) ([]*indexRecord, kv.Key, bool, error) {
+	w.idxRecords = w.idxRecords[:0]
+	start, end := handleRange.startKey, handleRange.excludedEndKey()
+	batchCnt := w.batchCnt * copReadBatchFactor
+	return fetchRowsFromCop(w.ctx, w.copCtx, start, end, txn.StartTS(), w.idxRecords, batchCnt)
+}
+
+// fetchRowsFromCop sends a coprocessor request and fetches the first batchCnt rows.
+func fetchRowsFromCop(ctx context.Context, copCtx *copContext, startKey, endKey kv.Key, startTS uint64,
+	buf []*indexRecord, batchCnt int) ([]*indexRecord, kv.Key, bool, error) {
+	srcResult, err := copCtx.buildTableScan(ctx, startTS, startKey, endKey)
+	if err != nil {
+		return nil, nil, false, errors.Trace(err)
+	}
+	var done bool
+	buf, done, err = copCtx.fetchTableScanResult(ctx, srcResult, buf, batchCnt)
+	nextKey := endKey
+	if !done {
+		lastHandle := buf[len(buf)-1].handle
+		prefix := tablecodec.GenTableRecordPrefix(copCtx.tblInfo.ID)
+		nextKey = tablecodec.EncodeRecordKey(prefix, lastHandle).Next()
+	}
+	return buf, nextKey, done, err
+}
+
+type copContext struct {
+	tblInfo  *model.TableInfo
+	idxInfo  *model.IndexInfo
+	pkInfo   *model.IndexInfo
+	colInfos []*model.ColumnInfo
+	fieldTps []*types.FieldType
+	sessCtx  sessionctx.Context
+
+	srcChunk *chunk.Chunk
+}
+
+func newCopContext(tblInfo *model.TableInfo, idxInfo *model.IndexInfo, sessCtx sessionctx.Context) *copContext {
+	colInfos := make([]*model.ColumnInfo, 0, len(idxInfo.Columns))
+	fieldTps := make([]*types.FieldType, 0, len(idxInfo.Columns))
+	for _, idxCol := range idxInfo.Columns {
+		c := tblInfo.Columns[idxCol.Offset]
+		colInfos = append(colInfos, c)
+		fieldTps = append(fieldTps, &c.FieldType)
+	}
+
+	pkColInfos, pkFieldTps, pkInfo := buildHandleColInfoAndFieldTypes(tblInfo)
+	colInfos = append(colInfos, pkColInfos...)
+	fieldTps = append(fieldTps, pkFieldTps...)
+
+	copCtx := &copContext{
+		tblInfo:  tblInfo,
+		idxInfo:  idxInfo,
+		pkInfo:   pkInfo,
+		colInfos: colInfos,
+		fieldTps: fieldTps,
+		sessCtx:  sessCtx,
+		srcChunk: chunk.NewChunkWithCapacity(fieldTps, variable.DefMaxChunkSize),
+	}
+	return copCtx
+}
+
+func (c *copContext) buildTableScan(ctx context.Context, startTS uint64, start, end kv.Key) (distsql.SelectResult, error) {
```

Comment:
It looks very like https://github.com/pingcap/tidb/blob/706a646c9f756213238c0f380e2aaa513eab07db/ddl/reorg.go#L438, could we combine those two functions?

Reply:
Sure, let's clean these up in another PR.